### PR TITLE
Use ruff as linter and code formatter

### PR DESCRIPTION
**Pull request by tkrabel**
Ruff is incredibly fast (running it on the pylsp code base is INSTANT; static code analysis in our pipeline now only takes 21s!) and a [drop-in replacement](https://docs.astral.sh/ruff/rules/#pyflakes-f) for pylint, pycodestyle, isort, and black.

This PR suggests the following:
1. Use ruff as a replacement for `pylint` and `pycodestyle`. This simplifies the devloop (no need to run two commands anymore) and makes it much faster (ruff is REALLY fast and offers an autofix option, which I used already to address 10+ issues).
2. Use ruff as a replacement for `black`. Since we use ruff as a linter, using it as a formatter keeps the devloop simple. The diff to black is minimal: only 5 function definitions we're indented differently. The rest was the same.
3. Use ruff for sorting imports. We don't have automated import sorting yet, and this would address issues like https://github.com/python-lsp/python-lsp-server/pull/499#discussion_r1435370352

### CODE REVIEW

**Comment by tkrabel** [comment]
Addresses https://github.com/python-lsp/python-lsp-server/pull/499#discussion_r1435370352

File: ruff.toml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+# Same as Black.
+line-length = 88
+indent-width = 4
+
+# Assume Python 3.8
+target-version = "py38"
+
+[lint]
+select = ["E", "F", "W", "C", "I"]
```
Comment (tkrabel) [inline ruff.toml:40]: `"I"` is for isort functionality

**Comment by tkrabel** [comment]
@ccordoba12 this PR is my Christmas present for you 😄. No more checking import order! Ruff takes care of that for you now 🎄 🎁 . Happy holidays! :)

File: setup.cfg
Comment (tkrabel) [inline setup.cfg:1]: Removed this since we're not using `pycodestyle` anymore. `ruff` takes care of that for us now.

**Comment by tkrabel-db** [comment]
@krassowski @ccordoba12 don't want to look impatient, but I'd love to get this in for all the contributors, making any lint/code style checks blazingly fast! :)

File: ruff.toml (hunk 1 above)
Comment (doolio) [reply ruff.toml:40]: Why not use the existing `pyproject.toml` rather than introduce a new config file in `ruff.toml`? 

Edit: cc @ccordoba12

**Comment by ccordoba12** [comment]
> this PR is my Christmas present for you 😄. No more checking import order! Ruff takes care of that for you now 🎄 🎁 . Happy holidays! :)

Thanks! This was a nice present indeed and sorry for answering until now but I was taking advantage of the holidays to finish several PRs in Spyder.

> don't want to look impatient, but I'd love to get this in for all the contributors, making any lint/code style checks blazingly fast! :)

Totally agree, so I'm going to merge your work now.

**Review by ccordoba12 — APPROVED** [review]
Thanks @tkrabel!

File: setup.cfg
Comment (doolio) [reply setup.cfg:1]: Can't we also remove `.pylintrc` as well?

cc: @ccordoba12

Comment (ccordoba12) [reply setup.cfg:1]: Yep, that's a good idea. I'll do that in a follow up PR.

Comment (doolio) [reply setup.cfg:1]: I submitted a PR for consideration for this just now.

**Comment by Shane-XB-Qian** [comment]
this maybe conflicted with ruff plugin? since seems that didnot use latest ruff for now.
why not just use that ruff plugin if like to use ruff linter and formatter?

--
shane.xb.qian

**Comment by doolio** [comment]
This PR was to use `ruff` to lint/format the `python-lsp-server` project not as a way for users to use `ruff` as a linter/formatter for their own projects **via** the `python-lsp-server`. For the latter users should use the `python-lsp-ruff` plugin as suggested in the README.

**Comment by Shane-XB-Qian** [comment]
ok. but since when ruff supported py project.toml instead of ruff.toml?
// I did not check the change detail, just roughly read this in the mail.

--
shane.xb.qian

**Comment by doolio** [comment]
> ok. but since when ruff supported py project.toml instead of ruff.toml?

I think from the very beginning. See [here](https://docs.astral.sh/ruff/configuration/).

Edit: Specifially

> If Ruff detects multiple configuration files in the same directory, the .ruff.toml file will take precedence over the ruff.toml file, and the ruff.toml file will take precedence over the pyproject.toml file.

**Pull request merged**